metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -44,47 +112,374 @@ class PackageProvider(private val buildTargets: BuildTargets) {
       }
     }
 
-    if (
-      path.isScalaOrJava && !path.isScalaScript && !path.isJarFileSystem &&
-      path.toFile.length() == 0
-    ) {
-      buildTargets
-        .inverseSourceItem(path)
-        .map(path.toRelative)
-        .flatMap(relativePath => Option(relativePath.toNIO.getParent))
-        .flatMap { parent =>
-          val pathIterator = parent.iterator().asScala
-          if (path.filename == "package.scala") {
-            packageObjectStatement(pathIterator)
-          } else {
-            val packageName = parent
-              .iterator()
-              .asScala
-              .map(p => wrap(p.toString()))
-              .mkString(".")
-            val text =
-              if (path.isScala) s"package $packageName\n\n@@"
-              else s"package $packageName;\n\n@@"
-            Some(NewFileTemplate(text))
+    for {
+      packageParts <- Option.when(
+        path.isScalaOrJava && !path.isJarFileSystem &&
+          !path.isScalaScript && path.toFile.length() == 0
+      )(deducePkgParts(path))
+      newFileTemplate <-
+        if (path.filename == "package.scala")
+          packageObjectStatement(packageParts)
+        else {
+          val packageName = packageParts.mkString(".")
+          val text =
+            if (path.isScala) s"package $packageName\n\n@@"
+            else s"package $packageName;\n\n@@"
+          Some(NewFileTemplate(text))
+        }
+      if packageParts.size > 0
+    } yield newFileTemplate
+  }
+
+  private def calcMovedFileEdits(
+      tree: Tree,
+      oldPath: AbsolutePath,
+      newPkgParts: Vector[String],
+  ): WorkspaceEdit = {
+    val pkgs = findPkgs(tree)
+
+    val oldPkgSplit: Vector[Vector[String]] =
+      pkgs.map(p => extractNames(p.ref))
+
+    val optPkgObj = tree.collect { case p: Pkg.Object => p }.headOption
+
+    // edit of package stament
+    val pkgStmtParts =
+      if (optPkgObj.isDefined) newPkgParts.dropRight(1) else newPkgParts
+    val pkgEdit = calcPkgEdit(oldPkgSplit, pkgStmtParts, oldPath, pkgs)
+
+    // edit of package object name if it exists in the file
+    val pkgObjEdit = for {
+      pkgObjStmt <- optPkgObj
+      lastPkg <- newPkgParts.lastOption
+      edit = mkWorkspaceEdit(oldPath, lastPkg, pkgObjStmt.name.pos.toLsp)
+      if pkgObjStmt.name.value != lastPkg
+    } yield edit
+
+    (List(pkgEdit) ++ pkgObjEdit).mergeChanges
+  }
+
+  private def calcRefsEdits(
+      movedTree: Tree,
+      oldPath: AbsolutePath,
+      oldPkgParts: Vector[String],
+      newPkgParts: Vector[String],
+      refProvider: ReferenceProvider,
+      optDirChange: Option[DirChange],
+  ) = {
+    val newPkgName = newPkgParts.mkString(".")
+    val decls = findTopLevelDecls(movedTree)
+    val oldUri = oldPath.toURI.toString
+    val edits = decls.view
+      .flatMap(decl => findRefs(decl, oldUri, refProvider).map((decl, _)))
+      .view
+      .groupMap(_._2.getUri())(_._1)
```

Comment:
I'm not a fan of accessing tuple members by index of field, could you rework it to `case ()` or create a case class and used it instead of tuple?

Reply:
Me neither. I can rework it if the following rewrite is more comprehensible:
``` scala
.groupMap { case (_, location) =>
    location.getUri()
 } { case (decl, _) =>
    decl
 }
```
Spawning one-shot case classes does not seem to be a proper way to reduce code complexity

Reply:
I think the case version is more readable though.

Reply:
Fixed